lib/Ebook.php

```diff
@@ -797,23 +1535,286 @@ private function NullIfEmpty($elements): ?string{
+			values (?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?,
+				?)
+		', [$this->Identifier, $this->WwwFilesystemPath, $this->RepoFilesystemPath, $this->KindleCoverUrl, $this->EpubUrl,
+				$this->AdvancedEpubUrl, $this->KepubUrl, $this->Azw3Url, $this->DistCoverUrl, $this->Title,
+				$this->FullTitle, $this->AlternateTitle, $this->Description, $this->LongDescription,
+				$this->Language, $this->WordCount, $this->ReadingEase, $this->GitHubUrl, $this->WikipediaUrl,
+				$this->EbookCreated, $this->EbookUpdated, $this->TextSinglePageByteCount, $this->IndexableText]);
+
+		$this->EbookId = Db::GetLastInsertedId();
+
+		$this->InsertTags();
+		$this->InsertLocSubjects();
+		$this->InsertCollectionMemberships();
+		$this->InsertGitCommits();
+		$this->InsertSources();
+		$this->InsertContributors();
+		$this->InsertTocEntries();
+	}
+
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Save(): void{
+		$this->Validate();
+
+		$this->InsertTagStrings();
+		$this->InsertLocSubjectStrings();
+		$this->InsertCollections();
+
+		Db::Query('
+			UPDATE Ebooks
+			set
+			Identifier = ?,
+			WwwFilesystemPath = ?,
+			RepoFilesystemPath = ?,
+			KindleCoverUrl = ?,
+			EpubUrl = ?,
+			AdvancedEpubUrl = ?,
+			KepubUrl = ?,
+			Azw3Url = ?,
+			DistCoverUrl = ?,
+			Title = ?,
+			FullTitle = ?,
+			AlternateTitle = ?,
+			Description = ?,
+			LongDescription = ?,
+			Language = ?,
+			WordCount = ?,
+			ReadingEase = ?,
+			GitHubUrl = ?,
+			WikipediaUrl = ?,
+			EbookCreated = ?,
+			EbookUpdated = ?,
+			TextSinglePageByteCount = ?,
+			IndexableText = ?
+			where
+			EbookId = ?
+		', [$this->Identifier, $this->WwwFilesystemPath, $this->RepoFilesystemPath, $this->KindleCoverUrl, $this->EpubUrl,
+				$this->AdvancedEpubUrl, $this->KepubUrl, $this->Azw3Url, $this->DistCoverUrl, $this->Title,
+				$this->FullTitle, $this->AlternateTitle, $this->Description, $this->LongDescription,
+				$this->Language, $this->WordCount, $this->ReadingEase, $this->GitHubUrl, $this->WikipediaUrl,
+				$this->EbookCreated, $this->EbookUpdated, $this->TextSinglePageByteCount, $this->IndexableText,
+				$this->EbookId]);
+
+		$this->DeleteTags();
+		$this->InsertTags();
+
+		$this->DeleteLocSubjects();
+		$this->InsertLocSubjects();
+
+		$this->DeleteCollectionMemberships();
+		$this->InsertCollectionMemberships();
+
+		$this->DeleteGitCommits();
+		$this->InsertGitCommits();
+
+		$this->DeleteSources();
+		$this->InsertSources();
+
+		$this->DeleteContributors();
+		$this->InsertContributors();
+
+		$this->DeleteTocEntries();
+		$this->InsertTocEntries();
+	}
+
+	private function DeleteTags(): void{
+		Db::Query('
+			DELETE from EbookTags
+			where EbookId = ?
+		', [$this->EbookId]
+		);
+	}
+
+	private function InsertTags(): void{
+		foreach($this->Tags as $tag){
+			Db::Query('
+				INSERT into EbookTags (EbookId, TagId)
```

Comment:
I added the check, but I wanted to point out that it's not possible to get a  `DuplicateDatabaseKeyException` from the code as it works today. `Ebook::Save()` calls:

```php  
$this->RemoveTags();  
$this->AddTags();  
```

so any previous entries in `EbookTags` are gone when `AddTags()` runs. I can elaborate on why I like this approach if it would help. Same approach for `LocSubjects`, `CollectionMemberships`, `GitCommits`, `Sources`, `Contributors`, and `TocEntries`.